### ISSUE TITLE
AxolotlGRPOTrainer still shuffles combined datasets even with curriculum_sampling flag enabled

### ISSUE THREAD

**Issue by sidmadala** [axolotl-ai-cloud/axolotl#2376]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Using the `curriculum_sampling` flag with the `AxolotlGRPOTrainer` should not shuffle my combined dataset when using GRPO post-training.

### Current behaviour

Enabling the `curriculum_sampling` flag in my `config.yaml` still results in a shuffled combined dataset even though I want my combined datasets to remain in their proper order since I am combining GRPO with CL. 

### Steps to reproduce

Refer to config.yaml and run training via:

### Preprocess Dataset
`python -m axolotl.cli.preprocess config.yaml `

### Run training on 8 GPUs with 1 reserved for vLLM
`accelerate launch --num_processes=7 -m axolotl.cli.train config.yaml --deepspeed /home/sidmadala/grpo-cl/deepspeed_configs/zero2.json`

### Config yaml

```yaml
base_model: Qwen/Qwen2.5-7B-Instruct

# Optionally keep these or modify
load_in_8bit: false
load_in_4bit: false
strict: false

# torch_compile: true

rl: grpo
trl:
  beta: 0.01
  max_completion_length: 800
  use_vllm: true
  reward_funcs:
    - hotpotqa_grpo.correctness_reward_func # 5
    - hotpotqa_grpo.citation1_reward_func # 2
    - hotpotqa_grpo.citation2_reward_func # 2
    # - hotpotqa_grpo.passage_mention_reward_func
    - hotpotqa_grpo.penalize_wrong_passages_reward_func
    - hotpotqa_grpo.xmlcount_reward_func # 0.5
    - hotpotqa_grpo.formatting_reward_func # 0.5
    - hotpotqa_grpo.unicode_reward_func
    - hotpotqa_grpo.length_reward_func
  vllm_gpu_memory_utilization: 0.7
  vllm_max_model_len: 2500
  num_generations: 7

chat_template: qwen_25

datasets:
  # Instead of `skrishna/gsm8k_only_answer`, your HotpotQA dataset
  # might be on HF or local. For example: 
  - path: sidmadala/disjoint_dataset1
    type: dataset_transform
  - path: sidmadala/disjoint_dataset2
    type: dataset_transform
  - path: sidmadala/disjoint_dataset3
    type: dataset_transform

shuffle_merged_datasets: false
curriculum_sampling: true  # NOTE: THIS IS WHAT IS FAILING!!!

# Optionally rename
dataset_prepared_path: /home/sidmadala/qa-analysis/last_run_prepared/
skip_prepare_dataset: false
val_set_size: 0.0

output_dir: /home/sidmadala/axolotl-artifacts/cl-outputs

# DataLoader + training details
dataloader_prefetch_factor: 2
dataloader_num_workers: 32
dataloader_pin_memory: true
gc_steps: 1

sequence_len: 2500
sample_packing: false
eval_sample_packing: false
pad_to_sequence_len: false

wandb_project: grpo-cl-test
wandb_entity:
wandb_name:

gradient_accumulation_steps: 6
micro_batch_size: 7  # Should match (num_generations / num_gpus)
num_epochs: 1

optimizer: adamw_torch_fused
lr_scheduler: constant_with_warmup
learning_rate: 1.0e-6
max_grad_norm: 1.0
weight_decay: 0

bf16: true
tf32: true

gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
flash_attention: true

logging_steps: 1
warmup_steps: 50
eval_strategy:
evals_per_epoch: 0
save_steps: 100
save_total_limit: 1
```

### Possible solution

We should modify `AxolotlGRPOTrainer` to have the same behavior enabled by #2235 for other trainers. However, this might be an issue with the underlying `GRPOTrainer` provided by `trl`, but I just wanted to bring it to the attention of the team at Axolotl.

NOTE: I have brought this issue up at [trl/issues/#2998](https://github.com/huggingface/trl/issues/2998#issue-2889747835)

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hey! 

Thanks for the report. Let's see what upstream trl does first.

**Comment by sidmadala** [comment]
It doesn't appear that Hugging Face would be interested in this change to their [RepeatRandomSampler's](https://github.com/huggingface/trl/blob/491921c1a4167e7c84429382470b0bb3158e66b0/trl/trainer/grpo_trainer.py#L78) `__iter__` method since prior posts indicate they prefer you just subclass and override the `_get_train_dataloader` method as described in [transformers/issues/11693](https://github.com/huggingface/transformers/issues/11693).

Unsure if there is any way to take the initiative and modify something within Axolotl so we can perform curriculum based training within GRPO. My current fix is within the referenced `trl` issue above, but I am unsure how to add that to axolotl without modifying the underlying `GRPOTrainer` implementation.

**Comment by NanoCode012** [comment]
Hey! Thanks for checking back.

In this case, you could override those 2 dataloader fn to return your custom `RepeatSampler` class. I looked a bit more and `curriculum_sampling` seems to only applied for Axo SFT Trainer, not pref tuning, so you would also need to pass it along via the `training_args`.

**Comment by sidmadala** [comment]
I think this problem has been solved as of [trl/pull/3334](https://github.com/huggingface/trl/pull/3334). We might just want to ensure that this behavior is also present in Axolotl when pulling the most recent version of trl and using the config above.
